File: src/main/java/de/silencio/activecraftcore/guis/profilemenu/inventory/GamemodeSwitcherProfile.java
package de.silencio.activecraftcore.guis.profilemenu.inventory;

import de.silencio.activecraftcore.guicreator.GuiCreator;
import de.silencio.activecraftcore.guicreator.GuiItem;
import de.silencio.activecraftcore.guis.profilemenu.ProfileMenu;
import de.silencio.activecraftcore.messages.ProfileMessages;
import lombok.EqualsAndHashCode;
import lombok.Getter;
import lombok.ToString;
import org.bukkit.Material;
import org.bukkit.entity.Player;

@Getter
@EqualsAndHashCode(callSuper = false)
@ToString
public class GamemodeSwitcherProfile extends GuiCreator {

    private final ProfileMenu profileMenu;
    private final Player player;
    private final Player target;

    private GuiItem survivalStack, creativeStack, spectatorStack, adventureStack, currentGamemodeStack;

    public GamemodeSwitcherProfile(ProfileMenu profileMenu) {
        super("gamemode_switcher_profile", 3, ProfileMessages.GamemodeSwitcherProfile.TITLE());
        this.profileMenu = profileMenu;
        this.player = profileMenu.getPlayer();
        this.target = profileMenu.getTarget();
        refresh();
        profileMenu.setGamemodeSwitcherProfile(this);
    }

    @Override
    public void refresh() {

        fillBackground(true);
        setBackItem(21);
        setCloseItem(22);
        setPlayerHead(profileMenu.getPlayerHead(), 4);


        setItem(creativeStack = new GuiItem(Material.GRASS_BLOCK)
                        .setDisplayName(ProfileMessages.GamemodeSwitcherProfile.CREATIVE(target))
                        .addClickListener(guiClickEvent -> player.performCommand("creative " + profileMenu.getTarget().getName())),
                11, player, "activecraft.gamemode.creative.others");


        survivalStack = new GuiItem(Material.IRON_SWORD)
                .setDisplayName(ProfileMessages.GamemodeSwitcherProfile.SURVIVAL(target))
                .addClickListener(guiClickEvent -> player.performCommand("survival " + profileMenu.getTarget().getName()));
        setItem(survivalStack,
                12, player, "activecraft.gamemode.survival.others");


        setItem(currentGamemodeStack = new GuiItem(Material.WHITE_STAINED_GLASS_PANE)
                .setDisplayName(switch (target.getGameMode()) {
                    case CREATIVE -> ProfileMessages.GamemodeSwitcherProfile.CURRENT_GAMEMODE_CREATIVE();
                    case SURVIVAL -> ProfileMessages.GamemodeSwitcherProfile.CURRENT_GAMEMODE_SURVIVAL();
                    case SPECTATOR -> ProfileMessages.GamemodeSwitcherProfile.CURRENT_GAMEMODE_SPECTATOR();
                    case ADVENTURE -> ProfileMessages.GamemodeSwitcherProfile.CURRENT_GAMEMODE_ADVENTURE();
                }),
                13);


        setItem(adventureStack = new GuiItem(Material.MAP)
                .setDisplayName(ProfileMessages.GamemodeSwitcherProfile.ADVENTURE(target))
                .addClickListener(guiClickEvent -> player.performCommand("adventure " + profileMenu.getTarget().getName())),
                14, player, "activecraft.gamemode.adventure.others");


        setItem(spectatorStack = new GuiItem(Material.ENDER_EYE)
                .setDisplayName(ProfileMessages.GamemodeSwitcherProfile.SPECTATOR(target))
                .addClickListener(guiClickEvent -> player.performCommand("spectator " + profileMenu.getTarget().getName())),
                15, player, "activecraft.gamemode.spectator.others");
    }
}

File: src/main/java/org/activecraft/activecraftcore/guis/profilemenu/inventory/GamemodeSwitcherProfile.kt
package org.activecraft.activecraftcore.guis.profilemenu.inventory

import org.activecraft.activecraftcore.guicreator.GuiClickEvent
import org.activecraft.activecraftcore.guicreator.GuiCreator
import org.activecraft.activecraftcore.guicreator.GuiItem
import org.activecraft.activecraftcore.guicreator.GuiNavigator.Companion.pushReplacement
import org.activecraft.activecraftcore.guis.profilemenu.ProfileMenu
import org.activecraft.activecraftcore.messages.MessageSupplier
import org.activecraft.activecraftcore.messages.PlayerMessageFormatter
import org.bukkit.Material
import org.bukkit.entity.Player
import java.util.*


class GamemodeSwitcherProfile(private val profileMenu: ProfileMenu) : GuiCreator(
    identifier = "gamemode_switcher_profile",
    rows = 3,
    title = profileMenu.messageSupplier.getMessage(PREFIX + "title")
) {
    private val player: Player = profileMenu.player
    private val target: Player = profileMenu.target
    private val msgFormatter = PlayerMessageFormatter(activeCraftCoreMessage!!)
    private val messageSupplier: MessageSupplier = profileMenu.messageSupplier
    private var survivalStack: GuiItem? = null
    private var creativeStack: GuiItem? = null
    private var spectatorStack: GuiItem? = null
    private var adventureStack: GuiItem? = null
    private var currentGamemodeStack: GuiItem? = null

    init {
        refresh()
    }

    private fun sendCommand(gamemode: String) {
        player.performCommand(gamemode + " " + profileMenu.target.name)
    }

    private fun createGamemodeItem(gamemode: String, material: Material): GuiItem {
        return GuiItem(material)
            .setDisplayName(messageSupplier.getFormatted(PREFIX + gamemode, msgFormatter))
            .addClickListener { guiClickEvent: GuiClickEvent ->
                sendCommand(gamemode)
                pushReplacement(player, guiClickEvent.gui!!.rebuild())
            }
    }

    private fun addGamemodeItem(gamemode: String, material: Material, slot: Int): GuiItem {
        val item = createGamemodeItem(gamemode, material)
        setItem(item, slot, player, "activecraft.gamemode.$gamemode.others")
        return item
    }

    override fun refresh() {
        msgFormatter.setTarget(profileMenu.profile)
        fillBackground(true)
        setItem(profileMenu.defaultGuiCloseItem, 21)
        setItem(profileMenu.defaultGuiBackItem, 22)
        setItem(profileMenu.playerHead, 4)
        creativeStack = addGamemodeItem("creative", Material.GRASS_BLOCK, 11)
        survivalStack = addGamemodeItem("survival", Material.IRON_SWORD, 12)
        setItem(
            GuiItem(Material.WHITE_STAINED_GLASS_PANE)
                .setDisplayName(
                    messageSupplier.getMessage(
                        PREFIX + "current-gamemode-" + target.gameMode.name.lowercase(
                            Locale.getDefault()
                        )
                    )
                ).also { currentGamemodeStack = it },
            13
        )
        adventureStack = addGamemodeItem("adventure", Material.MAP, 14)
        spectatorStack = addGamemodeItem("spectator", Material.ENDER_EYE, 15)
    }

    companion object {
        private const val PREFIX = "profile.gamemode-switcher-gui."
    }
}